internal/cloud/azure/application_identity.go

```diff
@@ -45,38 +46,53 @@ func (c *AzureConfig) NewServicePrincipalsClient(ctx context.Context) (ServicePr
 }
 
 type ApplicationIdentityConfig struct {
-	ID            string
-	Name          string
-	ApplicationID string
-	ClientID      string
-	ClientSecret  string
+	ID                     string
+	Name                   string
+	ApplicationID          string
+	ServicePrincipalID     string
+	ServicePrincipalSecret string
 }
 
 func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityConfig, appClient ApplicationClient, spClient ServicePrincipalsClient) error {
-	app, _, appErr := appClient.Create(ctx, msgraph.Application{
-		DisplayName: &config.Name,
-	})
-	if appErr != nil {
-		return appErr
+	// The ID is the service principal ID, so technically theres a chance that last time the App was made
+	// but the SP was not. We need to check if the Application was made and skip creation if it was.
+	if config.ApplicationID == "" {
+		app, _, appErr := appClient.Create(ctx, msgraph.Application{
+			DisplayName: &config.Name,
+		})
+		if appErr != nil {
+			return appErr
+		}
+		config.ApplicationID = *app.ID
+	}
+
+	// fetch the application to make sure it exists
+	app, _, err := appClient.Get(ctx, config.ApplicationID, odata.Query{})
+	if err != nil {
+		return fmt.Errorf("error retrieving Application with object ID %v: %w", config.ApplicationID, err)
 	}
-	config.ApplicationID = *app.ID
-	config.ID = *app.ID
 
 	sp, _, spErr := spClient.Create(ctx, msgraph.ServicePrincipal{
 		AppId: app.AppId,
 	})
 	if spErr != nil {
 		return spErr
 	}
-	config.ClientID = *sp.ID
+	config.ServicePrincipalID = *sp.ID
+	config.ID = *sp.ID
```

Comment:
Notice the switch-a-roo magic we are doing here.  The `id` of this resource is the **Service Principal ID**, not the Application ID (or client id).  This is so the `application_permission` for azure can work just like GCP and AWS and just do:

```
resource "mdxc_application_permission" "example" {
  application_identity_id = mdxc_application_identity.example.id
  permission = {...}
}
```

instead of:

```
resource "mdxc_application_permission" "example" {
  application_identity_id = mdxc_application_identity.example.azure_application_identity.service_principal_id
  permission = {...}
}
```

However, this does put us in a weird situation where the application *might* have been made in a previous run, but the SP creation failed, so the ID doesn't get updated or stored in state. The application will be orphaned in this scenario.  Not sure how to address this other than deleting the Application resource if the SP creation fails (which I'm not doing yet).

Reply:
@mclacore for context. This PR creates most of the resources in your file here: https://github.com/massdriver-cloud/terraform-modules/pull/33/files#r960073173

And that's on purpose. :) Chris has smashed this, and you the app side. 👏  Once this is merged that file will look a lot like my comment.